### PR TITLE
Fix tdd calculations in CalcInteriorSolarDistribution

### LINKED ISSUE #8356: Bug in TDD related calculations in CalcInteriorSolarDistribution
Issue overview
--------------
Three potential bugs in TDD related calculations in `CalcInteriorSolarDistribution`.

1. The reporting at the end of the function starts a new surface loop, but the CosInc [here](https://github.com/NREL/EnergyPlus/blob/f37f9762f1d98546443addfdbd0afd539b5ac8b9/src/EnergyPlus/SolarShading.cc#L8195) is not re-calculated. The value, CosInc(Surfnum), used here is always the last calculated value for the last surface in the previous loop.

2. There are multiple special calculations (that are supposed to happen) for TDD Diffusers in the calculation loop ([branch 1](https://github.com/NREL/EnergyPlus/blob/f37f9762f1d98546443addfdbd0afd539b5ac8b9/src/EnergyPlus/SolarShading.cc#L6387), [branch 2](https://github.com/NREL/EnergyPlus/blob/f37f9762f1d98546443addfdbd0afd539b5ac8b9/src/EnergyPlus/SolarShading.cc#L6815), [branch 3](https://github.com/NREL/EnergyPlus/blob/f37f9762f1d98546443addfdbd0afd539b5ac8b9/src/EnergyPlus/SolarShading.cc#L7015), [branch 4](https://github.com/NREL/EnergyPlus/blob/f37f9762f1d98546443addfdbd0afd539b5ac8b9/src/EnergyPlus/SolarShading.cc#L7046)), but Diffusers are not `ExtSolar` surface, and they get bypassed after this [line](https://github.com/NREL/EnergyPlus/blob/f37f9762f1d98546443addfdbd0afd539b5ac8b9/src/EnergyPlus/SolarShading.cc#L6334).

3. The branch [here](https://github.com/NREL/EnergyPlus/blob/f37f9762f1d98546443addfdbd0afd539b5ac8b9/src/EnergyPlus/SolarShading.cc#L6822) for TDD dome can never be reached as it's under an `else if` condition of `SurfWinWindowModelType(SurfNum) != WindowBSDFModel && SurfWinWindowModelType(SurfNum) != WindowEQLModel`, which also applies to TDD Domes.

**Defect file**: testfiles/DaylightingDeviceTubular.idf

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### LINKED ISSUE #8369: Incorrect local variable scope in CalcInteriorSolarDistribution
A potential bug in solar calculations in CalcInteriorSolarDistribution.

The reporting at the end of the function starts a new surface loop, but the [WinTransBmBmSolar](https://github.com/NREL/EnergyPlus/blob/649ab487ef39cf44c2a8703beaeb750d25e687bd/src/EnergyPlus/SolarShading.cc#L6206), WinTransBmDifSolar, [TBmBm + TBmDif](https://github.com/NREL/EnergyPlus/blob/649ab487ef39cf44c2a8703beaeb750d25e687bd/src/EnergyPlus/SolarShading.cc#L5998) are not re-calculated. The value, [WinTransBmBmSolar](https://github.com/NREL/EnergyPlus/blob/649ab487ef39cf44c2a8703beaeb750d25e687bd/src/EnergyPlus/SolarShading.cc#L8254), WinTransBmBmSolar, and [TBmBm + TBmDif](https://github.com/NREL/EnergyPlus/blob/649ab487ef39cf44c2a8703beaeb750d25e687bd/src/EnergyPlus/SolarShading.cc#L8260) used in the reporting loop are always the last calculated value for the last surface in the previous loop.

Defect file: testfiles/PurchAirWindowBlind_BlockBeamSolar.idf

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [x] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by xuanluo113**
Pull request overview
---------------------
 - Fixes #8356
 - Fixes #8369 
 - Code change explained in Issue descriptions.
 - Diffs in DaylightingDevicesTubular.idf and DaylightingDeviceShelf.idf are expected with the fix.

These are the changes for DaylightingDeviceTubular:
image

The changes look reasonable, but the heat gain/loss is wrong based on the other temperatures. Will post a new issue for that.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] ~If feature, test running new feature, try creative ways to break it~
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] ~Check any new function arguments for performance impacts~
 - [ ] ~Verify IDF naming conventions and styles, memos and notes and defaults~
 - [ ] ~If new idf included, locally check the err file and other outputs~

### CODE REVIEW

**Review by mjwitte — COMMENTED** [review]
@xuanluo113 The code changes look good.
Unit tests all pass when run serially.
I've added some plots showing the changes in the TDD.
But it seems like there should be diffs in the light shelf file. Stepping through the debugger, the code block is not getting used, but somehow the numbers aren't making it all the way to output.

**Comment by mjwitte** [comment]
@xuanluo113 Are you planning to look more closely at the lack of diff for the light shelf test file?

**Comment by xuanluo113** [comment]
@mjwitte I could do that by reporting related variables and add to the current regression test suite if necessary. However, in the meantime, I found some other similar bugs (non-tdd related) in this function (Issue #8369), which tested locally, may trigger more diffs (less in design day but more in year-round simulation reporting). Hope you can take another look at that issue to verify.

I plan to fix those together in this branch unless you suggest otherwise

**Comment by mjwitte** [comment]
@xuanariel Regarding the light shelf, I attempted to add output variables to that file and could not find any that were changing. If you have a test file that shows  changes with this branch, please share it.

**Comment by mjwitte** [comment]
@xuanluo113 (correct name this time!) Regarding the light shelf, I attempted to add output variables to that file and could not find any that were changing. If you have a test file that shows  changes with this branch, please share it.

**Comment by xuanluo113** [comment]
@mjwitte Thanks. 

1. I fixed and linked Issue #8369 here. 

2. For the light shelf, I took another look and found there are no diffs because the whole branches are calculating things never got used. The [problematic branch](https://github.com/NREL/EnergyPlus/blob/649ab487ef39cf44c2a8703beaeb750d25e687bd/src/EnergyPlus/SolarShading.cc#L6801) is calculating `DGZoneWin` and `DSZoneWin`. These are to add to `DSZone // Factor for sky diffuse solar radiation into a zone` and `DGZone // Factor for ground diffuse solar radiation into a zone` arrays.

I searched these arrays globally, they were used for calculating [QDforDaylight](https://github.com/NREL/EnergyPlus/blob/6015b8da62b3152569e0ed1361e897a3e054f0a3/src/EnergyPlus/HeatBalanceSurfaceManager.cc#L2644) (Diffuse solar radiation in a zone from sky and ground diffuse entering) back in 2007, but not anymore ever since. However, there still hang quite some blocks there to calculate these intermediate arrays.

Anyhow, I plan to leave this bug fixing branch as it is and switch back to window-struct2 (PR #8347) to fix the second issue. There are nested branches and variables related to those calculations and would be better to add the fix upon the refactoring branch. The branch window-struct2 is currently cleaned up to be reviewed next.

File: src/EnergyPlus/SolarShading.cc [hunk 1]
```diff
@@ -8190,6 +8160,7 @@ namespace SolarShading {
                         if (SurfWinOriginalClass(SurfNum) == SurfaceClass_TDD_Diffuser) {
                             PipeNum = SurfWinTDDPipeNum(SurfNum);
                             SurfNum2 = TDDPipe(PipeNum).Dome;
+                            CosInc = CosIncAng(TimeStep, HourOfDay, SurfNum2);
```
Comment (mjwitte) [inline src/EnergyPlus/SolarShading.cc:8163]: Seems like `CosInc` should be declared here with local scope (and also up in line 6360?). Possibly the same for some of these other variables. If you are planning that already in the next windowstruct branch, then it's ok here as-is.

File: src/EnergyPlus/SolarShading.cc [hunk 2]
```diff
@@ -6387,8 +6381,8 @@ namespace SolarShading {
                 if (Surface(SurfNum).Class == SurfaceClass_TDD_Dome || SurfWinOriginalClass(SurfNum) == SurfaceClass_TDD_Diffuser)
                     SunlitFracWithoutReveal(TimeStep, HourOfDay, SurfNum) = SunLitFract; // Frames/dividers not allowed
 
-                WinTransBmBmSolar = 0.0;
-                WinTransBmDifSolar = 0.0;
+                WinTransBmBmSolar(SurfNum) = 0.0;
+                WinTransBmDifSolar(SurfNum) = 0.0;
```
Comment (mjwitte) [inline src/EnergyPlus/SolarShading.cc:6385]: The full arrays for `WinTransBmBmSolar` and `WinTransBmDifSolar` are zeroed already in lines 6276:6277. Delete  one of these?

File: src/EnergyPlus/SolarShading.cc [hunk 3]
```diff
@@ -7252,21 +7223,21 @@ namespace SolarShading {
...
-                        WinTransBmDifSolar = TBmDif * SunLitFract * CosInc * Surface(SurfNum).Area * InOutProjSLFracMult; // m2
+                        WinTransBmBmSolar(SurfNum) = TBmBm * SunLitFract * CosInc * Surface(SurfNum).Area * InOutProjSLFracMult;   // m2
+                        WinTransBmDifSolar(SurfNum) = TBmDif * SunLitFract * CosInc * Surface(SurfNum).Area * InOutProjSLFracMult; // m2
                         WinTransBmSolar(SurfNum) += SurfWinOutsRevealDiffOntoGlazing(SurfNum) * NomDiffTrans * Surface(SurfNum).Area;
 
                         WinTransBmDifSolar += SurfWinOutsRevealDiffOntoGlazing(SurfNum) * NomDiffTrans * Surface(SurfNum).Area;
                     } else {
                         WinTransBmSolar(SurfNum) = 0.0;
-                        WinTransBmDifSolar = 0.0;
+                        WinTransBmDifSolar(SurfNum) = 0.0;
```
Comment (mjwitte) [inline src/EnergyPlus/SolarShading.cc:7233]: And these lines probably are not necessary if these have been zeroed up at the top of the loop, but it's hard to know for sure if the path leads directly here or if the reset to zero is necessary.

**Review by mjwitte — APPROVED** [review]
@xuanluo113 This looks good. A few comments that can be addressed here or in the next window-struct branch. Let me know. I've pulled in develop to let CI make another pass before merging.

**Comment by xuanluo113** [comment]
@mjwitte Thanks! These have been resolved or do not apply anymore in the `window-struct2` branch.  `CosInc` is declared locally. `WinTransBmBmSolar` and  `WinTransBmDifSolar` are declared at the top without initialization, and reset to zero only when necessary in the loop. The redundant resets are removed.

**Comment by mjwitte** [comment]
@xuanluo113 Thanks for clarifying. This will merge once CI finishes.

**Pull request merged**